### PR TITLE
feat(tabs): Creates Tabs component

### PR DESCRIPTION
**Pull request by diegonvs**

### CODE REVIEW

File: packages/clay-tabs/src/Item.tsx [hunk 1]
```diff
@@ -40,9 +40,9 @@ interface IProps extends Omit<React.HTMLAttributes<HTMLLIElement>, 'onClick'> {
 	>;
 
 	/**
-	 * Registers the `tabKey` for the Tab Item.
+	 * Registers the `tabkey` for the Tab Item.
 	 */
-	tabKey: number;
+	tabkey: number;
```
Comment (wincent) [inline packages/clay-tabs/src/Item.tsx]: Why the change from `tabKey` to `tabKey`? It is inconsistent with other props like `tabName` and `activeTabKey`?

File: packages/clay-tabs/src/Content.tsx [hunk 2]
```diff
@@ -29,13 +34,9 @@ const Content: React.FunctionComponent<IProps> = ({
...
 		<div className={classNames('tab-content', className)} {...otherProps}>
 			{React.Children.map(children, (child, index) => {
-				if (!React.isValidElement(child)) {
-					return null;
-				}
-
 				return React.cloneElement(child, {
 					...child.props,
-					active: activeTabKey === child.props.tabKey,
+					active: activeTabKey === child.props.tabkey,
```
Comment (wincent) [inline packages/clay-tabs/src/Content.tsx]: I feel a bit queasy about accessing `child.props` from the outside. The fact that the `TabPane` component takes `tabkey` as a prop but doesn't actually access it is another sign that this is a smell.

This feels like the kind of thing that the component should take care of instead of the user; ie. in the example below, having to provide a `tabkey` is just expanding with surface area of the API with another unique prop (like `tabName`), which is another thing for me to have to worry about and possibly get wrong. It feels a bit like a leaky abstraction; we're exposing what would ideally be internal-only plumbing:

```javascript
<ClayTabs>
	<ClayTabs.Item
		active={activeTabKeyValue == 0}
		onClick={onClick}
		tabName="Dummy1"
		tabkey={0}
	/>
	<ClayTabs.Item
		active={activeTabKeyValue == 1}
		data-testid="tabItem2"
		onClick={onClick}
		tabName="Dummy2"
		tabkey={1}
	/>
	<ClayTabs.Item
		active={activeTabKeyValue == 2}
		onClick={onClick}
		tabName="Dummy3"
		tabkey={2}
	/>
</ClayTabs>
```

it would be nice if we could take care of that. I have a feeling that could be made to work; did you explore that in the initial implementation?

File: packages/clay-tabs/src/Item.tsx [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+					{...itemElementProps}
+				>
+					{tabName}
+					{dropdown && (
+						<ClayIcon spritemap={spritemap} symbol="caret-bottom" />
+					)}
+				</button>
+			</>
+		);
+	}
```
Comment (wincent) [inline packages/clay-tabs/src/Item.tsx]: Should be no need to wrap in a fragment in either of these places because you're only returning a single element. Also, there is a lot of duplication between the `a` and `button` cases, can we factor that out a bit (for example, but spreading in the common props)?

File: packages/clay-tabs/src/Item.tsx [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+			className={classNames(
+				'nav-item',
+				{
+					dropdown,
+				},
+				className
+			)}
+			{...otherProps}
+		>
+			{tabName && (
```
Comment (wincent) [inline packages/clay-tabs/src/Item.tsx]: `tabName` is a required parameter, so why the guard here?

File: packages/clay-tabs/src/TabPane.tsx [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import React, {useEffect, useState} from 'react';
+
+interface ITabPaneProps extends React.HTMLAttributes<HTMLDivElement> {
+	/**
+	 * Flag to indicate if `active` classname should be applied
+	 */
+	active?: boolean;
+
+	/**
+	 * Flag to indicate if `fade` classname that applies an fading animation should be applied.
```
Comment (wincent) [inline packages/clay-tabs/src/TabPane.tsx]: "an fading" → "a fading"

File: packages/clay-tabs/src/TabPane.tsx [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+	/**
+	 * TabPane's key.
+	 */
+	tabkey: number;
+}
+
+const delay = (fn: Function) => {
+	return setTimeout(() => {
+		fn();
+	}, 100);
```
Comment (wincent) [inline packages/clay-tabs/src/TabPane.tsx]: Where did this value come from? Could it be moved into a named constant?

File: packages/clay-tabs/src/__tests__/index.tsx [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+};
+
+describe('ClayTabs', () => {
+	afterEach(cleanup);
+
+	it('renders', () => {
+		const {container} = render(<ClayTabs />);
+
+		expect(container).toMatchSnapshot();
+	});
```
Comment (wincent) [inline packages/clay-tabs/src/__tests__/index.tsx]: Don't we usually have blank lines between `it()` blocks? (Please ignore me if I'm mistaken about that)

File: packages/clay-tabs/src/Item.tsx (hunk 4 above)
Comment (diegonvs) [reply packages/clay-tabs/src/Item.tsx]: Yes, It's unnecessary. I was relying on Typescript to warn me about this. Removing 😅

File: packages/clay-tabs/src/TabPane.tsx (hunk 5 above)
Comment (diegonvs) [reply packages/clay-tabs/src/TabPane.tsx]: ✅

File: packages/clay-tabs/src/TabPane.tsx (hunk 6 above)
Comment (diegonvs) [reply packages/clay-tabs/src/TabPane.tsx]: Sure, I was minding about turn this delay a property. Makes sense?

/cc @bryceosterhaus @matuzalemsteles too

File: packages/clay-tabs/src/Item.tsx (hunk 3 above)
Comment (diegonvs) [reply packages/clay-tabs/src/Item.tsx]: Makes sense Greg, making changes.

File: packages/clay-tabs/src/Item.tsx [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+							/>
+						)}
+					</a>
+				</>
+			);
+		}
+
+		return (
+			<>
+				<button
```
Comment (matuzalemsteles) [inline packages/clay-tabs/src/Item.tsx]: Could we use `ClayButton` here?

File: packages/clay-tabs/src/Item.tsx [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+			onClick,
+			spritemap,
+			tabName,
+		}: IProps,
+		IProps
+	) => {
+		if (component === 'anchor') {
+			return (
+				<>
+					<a
```
Comment (matuzalemsteles) [inline packages/clay-tabs/src/Item.tsx]: Could we use `ClayLink` here?

File: packages/clay-tabs/src/Item.tsx [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+						aria-controls={tabName.toLowerCase()}
+						aria-disabled={!active}
+						aria-selected={!active}
+						className={classNames('nav-link', {
+							active,
+							disabled,
+						})}
+						data-testid={`${tabName.trim().toLowerCase()}TabItem`}
+						href={`#${tabName.trim().toLowerCase()}`}
+						id={`${tabName}Tab`}
```
Comment (matuzalemsteles) [inline packages/clay-tabs/src/Item.tsx]: Just `tabName` would be enough here.

File: packages/clay-tabs/src/Item.tsx [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+					<a
+						aria-controls={tabName.toLowerCase()}
+						aria-disabled={!active}
+						aria-selected={!active}
+						className={classNames('nav-link', {
+							active,
+							disabled,
+						})}
+						data-testid={`${tabName.trim().toLowerCase()}TabItem`}
+						href={`#${tabName.trim().toLowerCase()}`}
```
Comment (matuzalemsteles) [inline packages/clay-tabs/src/Item.tsx]: I think we should not impose that `href` is always an anchor, we should expose a `href` API instead of making it Implicit.

File: packages/clay-tabs/src/Item.tsx [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+						id={`${tabName}Tab`}
+						onClick={() => onClick && onClick(tabkey)}
+						ref={forwardRef}
+						role="tab"
+						tabIndex={disabled ? -1 : undefined}
+						{...itemElementProps}
+					>
+						{tabName}
+						{dropdown && (
```
Comment (matuzalemsteles) [inline packages/clay-tabs/src/Item.tsx]: I prefer this to be `showIcon` is more generic, `dropdown` brings another feeling.

File: packages/clay-tabs/src/Item.tsx [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+					id={`${tabName}Tab`}
+					onClick={() => onClick && onClick(tabkey)}
+					ref={forwardRef}
+					role="tab"
+					tabIndex={disabled ? -1 : undefined}
+					type="button"
+					{...itemElementProps}
+				>
+					{tabName}
+					{dropdown && (
```
Comment (matuzalemsteles) [inline packages/clay-tabs/src/Item.tsx]: The same here!

File: packages/clay-tabs/src/Item.tsx [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+		);
+	}
+);
+
+const Item: React.FunctionComponent<IProps> = ({
+	active = false,
+	component,
+	className,
+	disabled,
+	dropdown,
```
Comment (matuzalemsteles) [inline packages/clay-tabs/src/Item.tsx]: The same here!

File: packages/clay-tabs/src/TabPane.tsx (hunk 6 above)
Comment (matuzalemsteles) [reply packages/clay-tabs/src/TabPane.tsx]: I prefer to keep it away from the API, because this is just a fake time to cause the animation, developers don't need to know that.

File: packages/clay-tabs/src/__tests__/index.tsx [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+		fireEvent.click(getByTestId('oneTabItem'));
+		expect(onClick).toBeCalled();
+		expect(onClick).toBeCalledWith(1);
+
+		fireEvent.click(getByTestId('twoTabItem'));
+		expect(onClick).toBeCalled();
+		expect(onClick).toBeCalledWith(2);
+
+		expect(container).toMatchSnapshot();
```
Comment (matuzalemsteles) [inline packages/clay-tabs/src/__tests__/index.tsx]: It is necessary to run a `toMatchSnapshot` here, we will try to minimize the use of snapshot in our tests.

File: packages/clay-tabs/src/index.tsx [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+			role="tablist"
+			{...otherProps}
+		>
+			{children}
+		</ul>
+	);
+};
+
+ClayTabs.Content = Content;
+
```
Comment (matuzalemsteles) [inline packages/clay-tabs/src/index.tsx]: No space required here.

File: packages/clay-tabs/src/Item.tsx (hunk 1 above)
Comment (diegonvs) [reply packages/clay-tabs/src/Item.tsx]: I was getting errors with this property using camel case. See the following stacktrace:
```log

File: packages/clay-tabs/src/Content.tsx (hunk 2 above)
Comment (diegonvs) [reply packages/clay-tabs/src/Content.tsx]: Yes, Initially I implemented using the order of the `Item`s that matches with the order of `Content` but I wasn't able to use Dropdown using this logic.

Personally, I prefer this version of the [API](https://github.com/diegonvs/clay/blob/47526867448545c7372466cb55e58a1e0cab4f0f/packages/clay-tabs/stories/index.tsx), however, I don't know what is the number of the items that I need for creating a [dropdown](https://issues.liferay.com/browse/LEXI-664) and I was thinking about this API could be high-level. 

Check out the versions that I tried before here: https://github.com/diegonvs/clay/commits/tabs

File: packages/clay-tabs/src/__tests__/index.tsx (hunk 7 above)
Comment (diegonvs) [reply packages/clay-tabs/src/__tests__/index.tsx]: Yes, I guess that is a pattern here. 😅

Comment (diegonvs) [reply packages/clay-tabs/src/__tests__/index.tsx]: I gonna fix it. Would be awesome add an eslint rule for it

Comment (wincent) [reply packages/clay-tabs/src/__tests__/index.tsx]: AFAIK there is not en existing ESLint rule that we could use for this; the closest I could find was [padding-line-between-statements](https://eslint.org/docs/rules/padding-line-between-statements), which has a zillion configuration possibilities but none which would apply here.

We might have to make a custom rule for it. In any case, created a ticket for it [https://github.com/liferay/eslint-config-liferay/issues/65](here).

File: packages/clay-tabs/src/Item.tsx (hunk 10 above)
Comment (wincent) [reply packages/clay-tabs/src/Item.tsx]: How do you know it would be globally unique?

File: packages/clay-tabs/src/Item.tsx (hunk 1 above)
Comment (wincent) [reply packages/clay-tabs/src/Item.tsx]: I see, because you are including it with `otherProps` and passing those to the `div` element rendered by `TabPane`.

File: packages/clay-tabs/src/Item.tsx (hunk 10 above)
Comment (matuzalemsteles) [reply packages/clay-tabs/src/Item.tsx]: We don't need to know but interactively who is using it will want to change that, since Diego is changing we can have an `innerProps` and dev can add the id as needed, I find it odd to create it implicitly.

File: packages/clay-tabs/src/Item.tsx (hunk 1 above)
Comment (diegonvs) [reply packages/clay-tabs/src/Item.tsx]: I don't think that would be good to remove `otherProps`. I can just extract the value of `tabKey` and don't use. What do you think about it?

File: packages/clay-tabs/src/Item.tsx (hunk 8 above)
Comment (diegonvs) [reply packages/clay-tabs/src/Item.tsx]: As discussed internally, we're waiting for a component to resolve when to use Button and when use Anchor. This code review is outdated. Currently is using an `ElementTag` like implementation.

File: packages/clay-tabs/src/Item.tsx (hunk 9 above)
Comment (diegonvs) [reply packages/clay-tabs/src/Item.tsx]: As discussed internally, we're waiting for a component to resolve when to use Button and when use Anchor. This code review is outdated. Currently is using an `ElementTag` like implementation.

File: packages/clay-tabs/src/Item.tsx (hunk 11 above)
Comment (diegonvs) [reply packages/clay-tabs/src/Item.tsx]: Okay

File: packages/clay-tabs/src/Item.tsx (hunk 12 above)
Comment (diegonvs) [reply packages/clay-tabs/src/Item.tsx]: `dropdown` isn't just adding an icon, is adding a `.dropdown` class too.

File: packages/clay-tabs/src/Item.tsx (hunk 13 above)
Comment (diegonvs) [reply packages/clay-tabs/src/Item.tsx]: `dropdown` isn't just adding an icon, is adding a `.dropdown` class too.

File: packages/clay-tabs/src/Item.tsx (hunk 14 above)
Comment (diegonvs) [reply packages/clay-tabs/src/Item.tsx]: `dropdown` isn't just adding an icon, is adding a `.dropdown` class too.

File: packages/clay-tabs/src/__tests__/index.tsx (hunk 15 above)
Comment (diegonvs) [reply packages/clay-tabs/src/__tests__/index.tsx]: Nope. Removing.

File: packages/clay-tabs/src/index.tsx (hunk 16 above)
Comment (diegonvs) [reply packages/clay-tabs/src/index.tsx]: ✅

**Comment by diegonvs** [comment]
Hey folks, I pushed some things after your comments ;)

File: packages/clay-tabs/src/Item.tsx (hunk 1 above)
Comment (wincent) [reply packages/clay-tabs/src/Item.tsx]: 👍 Yes, that's exactly what I meant by "you should remove it from `otherProps`".

**Comment by matuzalemsteles** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by matuzalemsteles** [comment]
Closing in favor of #2317
